### PR TITLE
Fix/37 and engprd 668

### PR DESCRIPTION
**Pull request by imkarrer**
gradle-plugins/#37 In dev

Rel: https://elasticco.atlassian.net/browse/ENGPRD-668?focusedCommentId=151493

### CODE REVIEW

File: plugins/docker/base-image/src/main/resources/archive-packages.sh [hunk 1]
```diff
@@ -98,13 +99,20 @@ archive_apk_packages() {
 
   # Packages in the base image don't upgrade with apk upgrade because they are locked in /etc/apk/world so we need to
   # do it explicitly. URLS will point to the last version in the registry.
+  # Regex test cases and validation here
+  # https://regex101.com/r/b5FQIj/1
+  url_pattern=".*\/(.*)-((([0-9]+\.)+([0-9]+)|([0-9]+)|([0-9a-z_.]+))-(.*))?\.apk"
```
Comment (nordbergm) [inline plugins/docker/base-image/src/main/resources/archive-packages.sh:104]: I'm fairly certain we need the release as part of the version, so in the regex examples:
https://packages.wolfi.dev/os/aarch64/apk-tools-2.14.4-r0.apk => `2.14.4-r0`
https://packages.wolfi.dev/os/aarch64/ca-certificates-bundle-20240705-r0.apk => `20240705-r0`
https://packages.wolfi.dev/os/aarch64/ncurses-6.5_p20240629-r0.apk => `p20240629-r0`

There are many different versioning schemes used by Chainguard and simpler regex might be necessary without trying to match major.minor.patch as it's not always what's used.

Comment (nordbergm) [reply plugins/docker/base-image/src/main/resources/archive-packages.sh:104]: I wrote something like this which might be what we're looking for? https://regex101.com/r/Qd4UG7/1

```regex
(aarch64|x86_64)\/(.*)-([0-9].*).*\.apk
```

**Comment by imkarrer** [comment]
I am getting an error locally when running the IT that dont run in CI

`./gradlew :plugins:docker:base-image:integrationTest --tests "*DockerBaseImageMatrix"`

```
* What went wrong:
Execution failed for task ':dockerBaseImageLockfile'.
> CSV line from script not valid: OK: 20 MiB in 20 packages
```

Sure enough its the first line of the CSV
```
    OK: 20 MiB in 20 packages
    apk-tools,2.14.4-r0,,aarch64
    bash,5.2.21-r5,,aarch64
    busybox,1.36.1-r10,,aarch64
```

**Comment by imkarrer** [comment]
I am getting an error locally when running the IT that dont run in CI

`./gradlew :plugins:docker:base-image:integrationTest --tests "*DockerBaseImageMatrix"`

```
* What went wrong:
Execution failed for task ':dockerBaseImageLockfile'.
> CSV line from script not valid: OK: 20 MiB in 20 packages
```

Sure enough its the first line of the CSV
```
    OK: 20 MiB in 20 packages
    apk-tools,2.14.4-r0,,aarch64
    bash,5.2.21-r5,,aarch64
    busybox,1.36.1-r10,,aarch64
```

File: plugins/docker/base-image/src/main/java/co/elastic/gradle/dockerbase/DockerLockfileTask.java [hunk 2]
```diff
@@ -270,6 +270,7 @@ private void writeLockfile(ByteArrayOutputStream csvStream) throws IOException {
         }
         try (Reader reader = new StringReader(csvString)) {
             CSVParser parser = CSVParser.parse(reader, CSVFormat.DEFAULT);
+            System.out.println(csvString);
```
Comment (imkarrer) [inline plugins/docker/base-image/src/main/java/co/elastic/gradle/dockerbase/DockerLockfileTask.java:273]: delete this

File: plugins/docker/base-image/src/main/java/co/elastic/gradle/dockerbase/DockerBaseImageBuildPlugin.java [hunk 3]
```diff
@@ -262,7 +263,7 @@ protected static void addPackageAsDependency(Project target, BaseImageExtension
         final Map<String, String> dependencyNotation = Map.of(
                 "group", type + (
                         extension.getOSDistribution().get().equals(OSDistribution.WOLFI) ?
-                                "/" + Architecture.current().name().toLowerCase(Locale.ROOT)  :
+                                "/" + fromDockerName(pkg.architecture()).name().toLowerCase(Locale.ROOT) :
```
Comment (alpar-t) [inline plugins/docker/base-image/src/main/java/co/elastic/gradle/dockerbase/DockerBaseImageBuildPlugin.java:266]: I don't think this is the way to go. Architecture in the package name is optional and some distros have something like `all` in here for packages that are arch independent. The caller has the architecture that needs to go here, it should be passed in as a parameter.

File: plugins/docker/base-image/src/main/resources/archive-packages.sh [hunk 4]
```diff
@@ -142,6 +138,7 @@ archive_apk_packages() {
     echo "File $FILE not found!" >&2
     exit 1
   fi
+  arch=$PACKAGES_ARCH
```
Comment (alpar-t) [inline plugins/docker/base-image/src/main/resources/archive-packages.sh:141]: The arch in the csv is meant to reflect the arch in the package name. Wolfi doesn't add the arch there, it uses different repos.

File: plugins/docker/base-image/src/main/java/co/elastic/gradle/dockerbase/DockerBaseImageBuildPlugin.java (hunk 3 above)
Comment (ascottElastic) [reply plugins/docker/base-image/src/main/java/co/elastic/gradle/dockerbase/DockerBaseImageBuildPlugin.java:266]: @alpar-t If this isn't the right approach then do we know what is?

File: plugins/docker/base-image/src/main/resources/archive-packages.sh (hunk 4 above)
Comment (imkarrer) [reply plugins/docker/base-image/src/main/resources/archive-packages.sh:141]: wolfi puts the arch in the url, but I suppose thats being pedantic.  Ill remove and we will tell @jowiho the absence of an arch for a set of packages in a lockfile is normal?

File: plugins/docker/base-image/src/main/java/co/elastic/gradle/dockerbase/DockerBaseImageBuildPlugin.java (hunk 3 above)
Comment (imkarrer) [reply plugins/docker/base-image/src/main/java/co/elastic/gradle/dockerbase/DockerBaseImageBuildPlugin.java:266]: Ill dig some, but if we are not using the current arch, and we are not deriving from the packages, from where are we deriving it?  I suppose we are looping through arches in the calling context?

**Review by alpar-t — APPROVED** [review]

**Pull request merged**